### PR TITLE
Laboratory work #1, Garanina Maria - 19FPL1

### PR DESCRIPTION
**Pull request by mgaranina**

### CODE REVIEW

**Comment by dmitry-uraev** [comment]
The deadline is not that far - consider pushing changes so that we can review them.

**Comment by mgaranina** [comment]
"in get_adjacent_words function it is necessary to use get_concordance function"

could you plese clarify this? what does it mean exactly?

**Comment by dmitry-uraev** [comment]
"in get_adjacent_words function it is necessary to use get_concordance function"

could you plese clarify this? what does it mean exactly?

Originally posted by @mgaranina in #21 (comment)

You have to call implemented get_concordance function inside get_adjacent_words function to get concordances and then work with this concordances to find adjacent words

e.g

def get_adjacent_words(...):
concordances = get_concordance(...)
# get adjacent words out of concordances

File: lab_1/get_concordance_test.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def test_big_text_get_concordance_term(self):
         """
         Checks if a context for a given term can be found properly
         """
-        text = read_from_file('lab_1/data.txt')
+        text = read_from_file('data.txt')
```
Comment (dmitry-uraev) [inline lab_1/get_concordance_test.py]: do not change this paths while pushing to repository

File: lab_1/get_concordance_test.py [hunk 2]
```diff
@@ -113,7 +113,7 @@ def test_get_concordance_several_contexts_big_text(self):
         """
         Checks if contexts for a given term can be found in real text properly
         """
-        text = read_from_file('lab_1/data.txt')
+        text = read_from_file('data.txt')
```
Comment (dmitry-uraev) [inline lab_1/get_concordance_test.py]: same as above here

File: lab_1/get_concordance_test.py (hunk 1 above)
Comment (dmitry-uraev) [reply lab_1/get_concordance_test.py]: and do not change tests at all, or keep this changes only on your computer

File: lab_1/main.py [hunk 3]
```diff
@@ -12,7 +12,19 @@ def tokenize(text: str) -> list:
     e.g. text = 'The weather is sunny, the man is happy.'
     --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     """
-    pass
+    token_list = []
+    # punctuation = '!"#$%&\'()*+,-./:;<=>?@[\]^_`{|}~.'
+    try:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: try-except solution is applicable for one-line code where we expect errors, not whole function implementation. Consider writing conditions or other checks.

File: lab_1/main.py [hunk 4]
```diff
@@ -25,7 +37,16 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    tokens1 = []
+    if not isinstance(tokens, list):
+        pass
```
Comment (dmitry-uraev) [inline lab_1/main.py]: this solution will not work, better return []

File: lab_1/main.py [hunk 5]
```diff
@@ -25,7 +37,16 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
+    elif not stop_words or not isinstance(stop_words, list):
+        tokens1 = tokens
+    else:
+        for i in tokens:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: better avoid nesting, especially when it can be avoided

Comment (dmitry-uraev) [reply lab_1/main.py]: just check input params first as you have done with `if not isinstance(tokens, list):` and then implement main logic without nesting

File: lab_1/main.py [hunk 6]
```diff
@@ -36,7 +57,14 @@ def calculate_frequencies(tokens: list) -> dict:
     e.g. tokens = ['weather', 'sunny', 'man', 'happy']
     --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
     """
-    pass
+    freq_dict = {}
+    if not isinstance(tokens, list):
+        pass
```
Comment (dmitry-uraev) [inline lab_1/main.py]: `return {}` check for explanation above

File: lab_1/main.py [hunk 7]
```diff
@@ -36,7 +57,14 @@ def calculate_frequencies(tokens: list) -> dict:
...
+    else:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: avoid nesting

File: lab_1/main.py [hunk 8]
```diff
@@ -49,7 +77,22 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
     top_n = 1
     --> ['happy']
     """
-    pass
+    word_list = []
+    freq_list = []
+    top_n_words_list = []
+    try:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: try-except solution is applicable for one-line code where we expect errors, not whole function implementation. Consider writing conditions or other checks.

File: lab_1/main.py [hunk 9]
```diff
@@ -69,7 +112,22 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
     right_context_size = 3
     --> [['man', 'is', 'happy', 'the', 'dog', 'is'], ['dog', 'is', 'happy', 'but', 'the', 'cat']]
     """
-    pass
+    concordance = []
+    if (not isinstance(tokens, list) or not isinstance(word, str) or not word.isalpha() or
+            type(left_context_size) is not int or type(right_context_size) is not int or
+            (left_context_size < 1 and right_context_size < 1)):
+        pass
```
Comment (dmitry-uraev) [inline lab_1/main.py]: specify return statement here.  Check for explanation above

File: lab_1/main.py [hunk 10]
```diff
@@ -69,7 +112,22 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
...
+    else:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: avoid nesting. Check for explanation above

File: lab_1/tokenize_test.py [hunk 11]
```diff
@@ -59,7 +59,7 @@ def test_tokenize_big_text_case(self):
         """
         Tokenize big input text scenario
         """
-        text = read_from_file('lab_1/tokens.txt')
+        text = read_from_file('tokens.txt')
```
Comment (dmitry-uraev) [inline lab_1/tokenize_test.py]: keep this paths as specified for repository on GitHub

File: lab_1/tokenize_test.py [hunk 12]
```diff
@@ -69,7 +69,7 @@ def test_tokenize_big_text_length_equal(self):
         """
         Tokenize big input text and assert equal
         """
-        text = read_from_file('lab_1/tokens.txt')
+        text = read_from_file('tokens.txt')
```
Comment (dmitry-uraev) [inline lab_1/tokenize_test.py]: same here

**Comment by dmitry-uraev** [comment]
Overall comment:
Nice job! 
1. try to fix nesting issues.
2. specify `return` statements, not `pass`
3. do not touch paths in tests. Return path to default condition.
4. Start implementing start.py file.

**Comment by mgaranina** [comment]
didn't really understand your comments about nesting :( better not to use else statements? 

fixed the rest.

**Comment by mgaranina** [comment]
shouldn't the article 'a' be removed from tokens as a stop word? there is no 'a' in stop_words.txt

and what if in tokens i have strings that consist of only one letter? is it ok?

**Comment by dmitry-uraev** [comment]
Write via telegram. I'll explain about nesting and other questions

File: lab_1/main.py [hunk 13]
```diff
@@ -25,7 +36,13 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    if not isinstance(tokens, list):
+        return []
+    elif not isinstance(stop_words, list) or not stop_words:
+        return tokens
+    else:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: this return is not needed. just write `tokens1 = [i for i in tokens if i not in stop_words]
        return tokens1`

File: lab_1/main.py (hunk 7 above)
Comment (dmitry-uraev) [reply lab_1/main.py]: same here

**Review by dmitry-uraev — COMMENTED** [review]
Overall comment:
1. fix else statements - they are not needed here.
2. 'a' not in stop_words, but it should. Just let it be so

File: lab_1/main.py [hunk 14]
```diff
@@ -25,7 +35,12 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    if not isinstance(tokens, list):
+        return []
+    elif not isinstance(stop_words, list) or not stop_words:
+        return tokens
+    tokens1 = [i for i in tokens if i not in stop_words]
+    return tokens1
```
Comment (dmitry-uraev) [inline lab_1/main.py]: you can return here `return [i for i in tokens if i not in stop_words]` directly

File: lab_1/main.py [hunk 15]
```diff
@@ -25,7 +35,12 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    if not isinstance(tokens, list):
+        return []
+    elif not isinstance(stop_words, list) or not stop_words:
```
Comment (dmitry-uraev) [inline lab_1/main.py]: this expression can be moved in `if not isinstance(tokens, list):` expression

File: lab_1/main.py [hunk 16]
```diff
@@ -36,7 +51,13 @@ def calculate_frequencies(tokens: list) -> dict:
...
     --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
     """
-    pass
+    if not isinstance(tokens, list) or not all(isinstance(i, str) for i in tokens):
+        return {}
+    freq_dict = {}
+    for i in tokens:
+        if i not in freq_dict:
+            freq_dict[i] = tokens.count(i)
+    return freq_dict
```
Comment (dmitry-uraev) [inline lab_1/main.py]: good

File: lab_1/main.py [hunk 17]
```diff
@@ -49,7 +70,21 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
...
+    for k, v in freq_dict.items():
+        word_list.append(k)
+        freq_list.append(v)
+    if top_n > len(word_list):
+        top_n = len(word_list)
+    for i in range(top_n):
+        max_num_index = freq_list.index(max(freq_list))
+        freq_list[max_num_index] = 0
+        top_n_words_list.append(word_list[max_num_index])
+    return top_n_words_list
```
Comment (dmitry-uraev) [inline lab_1/main.py]: good

File: lab_1/main.py [hunk 18]
```diff
@@ -12,7 +12,17 @@ def tokenize(text: str) -> list:
...
+        return []
+    token_list = []
+    for i in text.split():
+        if not i.isalpha():
+            for s in i:
+                if not s.isalpha():
+                    i = i.replace(s, '')
+        if i.isalpha():
+            token_list.append(i.lower())
+    return token_list
```
Comment (dmitry-uraev) [inline lab_1/main.py]: can be simplified, but acceptable

File: lab_1/main.py [hunk 19]
```diff
@@ -69,7 +104,21 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
...
+    if left_context_size < 0:
+        left_context_size = 0
+    elif right_context_size < 0:
+        right_context_size = 0
+    for i in tokens:
+        if i == word:
+            word_index = tokens.index(i)
+            concordance.append(tokens[word_index - left_context_size: word_index + right_context_size + 1])
+            tokens.remove(i)
+    return concordance
```
Comment (dmitry-uraev) [inline lab_1/main.py]: good

File: lab_1/main.py [hunk 20]
```diff
@@ -87,7 +136,16 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
+    concordance = get_concordance(tokens, word, left_n, right_n)
+    adjacent_words_list = []
+    for i in concordance:
+        if left_n <= 0:
+            adjacent_words_list.append([i[-1]])
+        elif right_n <= 0:
+            adjacent_words_list.append([i[0]])
+        else:
+            adjacent_words_list.append([i[0], i[-1]])
+    return adjacent_words_list
```
Comment (dmitry-uraev) [inline lab_1/main.py]: good

File: lab_1/main.py [hunk 21]
```diff
@@ -125,4 +186,19 @@ def sort_concordance(tokens: list, word: str, left_context_size: int, right_cont
...
+        word_index = -1
+    sorted_concordance = []
+    words_list = [i[word_index] for i in concordance]
+    for i in sorted(words_list):
+        for l in concordance:
+            if i == l[word_index]:
+                sorted_concordance.append(l)
+                concordance.remove(l)
+                break
+    return sorted_concordance
```
Comment (dmitry-uraev) [inline lab_1/main.py]: good

File: lab_1/start.py [hunk 22]
```diff
@@ -2,6 +2,8 @@
 Concordance implementation starter
 """
 
+# some changes there
+
 from main import read_from_file
 import os
```
Comment (dmitry-uraev) [inline lab_1/start.py]: please, start implementing start.py

**Comment by mgaranina** [comment]
get_adjacent_words and sort_concordance don't work in start.py, empty lists are returned. don't know why :(

**Comment by dmitry-uraev** [comment]
Fix start.py issue and we are done here.

**Comment by mgaranina** [comment]
everything seems to work fine now

**Comment by dmitry-uraev** [comment]
Well done!

**Comment by dmitry-uraev** [comment]
done

**Pull request closed without merging**